## Analysis of Hacker News Posts

[Hacker News](https://news.ycombinator.com/) is a site started by Y combinator, where user-submitted stories known as posts are voted and commented upon.

The project aims to analyse the posts and comments on Hacker News with the objective to answer the following queestions.

1. Do Ask HN or Show HN receive more comment on average?
2. Do post created at a certain time receive more comment on average?

The result of this analysis will be posted and the conclusion will be found in the texts that follows.

In [1]:
opened_file = open("hacker_news.csv")
from csv import reader
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]

print(headers)
print("\n")
print(hn[:5], end = "")

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]

In [2]:
print(len(hn))

20100


In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [4]:
print(ask_posts[:2])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']]


In [5]:
total_ask_comments = 0
for row in ask_posts:
    comment = int(row[4])
    total_ask_comments+=comment
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

total_show_comments = []
for row in show_posts:
    show_comment = int(row[4])
    total_show_comments.append(show_comment)

avg_show_comments = sum(total_show_comments)/ len(total_show_comments)

print(avg_show_comments)
    

14.038417431192661
10.31669535283993


The average number of comments on `ASK HN` is 14 while the average number of coomments on `SHOW HN` is 10.3. 

To answer the first part of our analysis, `ASK HN` receives more comment on average than `SHOW HN` posts.

In [6]:
import datetime as dt

result_list = []

for row in ask_posts:
    time_created = row[6]
    comment = int(row[4])
    result_list.append([time_created, comment])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    posts_date = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    posts_hour = posts_date.strftime("%H")
    comment = row[1]
    if posts_hour not in counts_by_hour:
        counts_by_hour[posts_hour] = 1
        comments_by_hour[posts_hour] = comment
    else:
        counts_by_hour[posts_hour] += 1
        comments_by_hour[posts_hour] += comment

print(comments_by_hour)
print(counts_by_hour)

{'20': 1722, '13': 1253, '14': 1416, '05': 464, '22': 479, '08': 492, '10': 793, '04': 337, '16': 1814, '02': 1381, '15': 4477, '21': 1745, '00': 447, '03': 421, '07': 267, '12': 687, '06': 397, '19': 1188, '18': 1439, '17': 1146, '23': 543, '11': 641, '01': 683, '09': 251}
{'20': 80, '13': 85, '14': 107, '05': 46, '22': 71, '08': 48, '10': 59, '04': 47, '16': 108, '02': 58, '15': 116, '21': 109, '00': 55, '03': 54, '07': 34, '12': 73, '06': 44, '19': 110, '18': 109, '17': 100, '23': 68, '11': 58, '01': 60, '09': 45}


In [7]:
print(result_list[:5])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [8]:
total_posts = 0
for post in counts_by_hour:
    sum_num = counts_by_hour[post]
    total_posts +=sum_num
print(total_posts)

1744


In [9]:
total_comments = 0
for post in comments_by_hour:
    sum_num = comments_by_hour[post]
    total_comments +=sum_num
print(total_comments)

24483


In [10]:
average_placeholder_for_comment = total_comments/total_posts
print(average_placeholder_for_comment)

14.038417431192661


In [11]:
avg_by_hour = []
for hour in comments_by_hour:
    avg = comments_by_hour[hour]/ counts_by_hour[hour]
    avg_by_hour.append([hour, avg])

print(avg_by_hour)

    
    
    

[['20', 21.525], ['13', 14.741176470588234], ['14', 13.233644859813085], ['05', 10.08695652173913], ['22', 6.746478873239437], ['08', 10.25], ['10', 13.440677966101696], ['04', 7.170212765957447], ['16', 16.796296296296298], ['02', 23.810344827586206], ['15', 38.5948275862069], ['21', 16.009174311926607], ['00', 8.127272727272727], ['03', 7.796296296296297], ['07', 7.852941176470588], ['12', 9.41095890410959], ['06', 9.022727272727273], ['19', 10.8], ['18', 13.20183486238532], ['17', 11.46], ['23', 7.985294117647059], ['11', 11.051724137931034], ['01', 11.383333333333333], ['09', 5.5777777777777775]]


In [12]:
swap_avg_by_hour = []
for element in avg_by_hour:
    swap_avg_by_hour.append([element[1], element[0]])
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap





[[21.525, '20'], [14.741176470588234, '13'], [13.233644859813085, '14'], [10.08695652173913, '05'], [6.746478873239437, '22'], [10.25, '08'], [13.440677966101696, '10'], [7.170212765957447, '04'], [16.796296296296298, '16'], [23.810344827586206, '02'], [38.5948275862069, '15'], [16.009174311926607, '21'], [8.127272727272727, '00'], [7.796296296296297, '03'], [7.852941176470588, '07'], [9.41095890410959, '12'], [9.022727272727273, '06'], [10.8, '19'], [13.20183486238532, '18'], [11.46, '17'], [7.985294117647059, '23'], [11.051724137931034, '11'], [11.383333333333333, '01'], [5.5777777777777775, '09']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [13]:
print("Top 5 Hours for Ask Posts Comments")

template = "{}: {:.2f} average comments per post"

for row in sorted_swap[:5]:
    hour_str = dt.datetime.strptime(row[1], "%H")
    hour = hour_str.strftime("%H:%M")
    comment = row[0]
    output = template.format(hour, comment)
    print(output)

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


In [14]:
fiften_hour = 38.59
twenty_one_hour = 16.01
total = fiften_hour + twenty_one_hour
percentage = fiften_hour/total * 100
"{:.2f}%".format(percentage)

'70.68%'

From above, we see that posts created at 15:00 which equates to 21:00 CEST receives more comments than any other hour. Based on our analysis, posts created at this hour would receive 70% more comments than posts created at 03:00 CEST.

## Conclusion

In this study, we analyzed the analyse the posts and comments on Hacker News with the objective to answer the following queestions.

Having cleaned our data, leaving us with only posts that have at least one comment, we found that on average, posts with `ASK HN` receive more comments that posts with starts with `SHOW HN`.

We also found that out of the posts that starts with `ASK HN`, the best time to post with high average number of comments is at 21:00 CEST.